server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java

```diff
@@ -402,6 +400,10 @@ private void extractItem(String path, XContentParser source) throws IOException
                 }
             }
 
+            private void addHash(String path, BytesRef value) {
+                hashes.add(new NameAndHash(new BytesRef(path), hash(value), hashes.size()));
```

Comment:
Do I understand correctly that the hash here depends on the order the fields get hashed? I wonder if we have issues if the fields get shuffled, i.e. `"foo": { "a": [1,2,3], "b": 10  }` and `"foo": { "b": 10, "a": [1,2,3] }` need to give the same hash.

Reply:
No, the hashes get sorted by the field name, so the hash is consistent, regardless of the ordering of fields in the source. Where ordering matters is within arrays. This is why in the `compare to` method in `NameAndHash`, we're primarily sorting by the name, and the order is just the tie breaker, which only affects arrays, as that's the only way you can have two `NameAndHash` instances with the same name. This is also tested in `org.elasticsearch.cluster.routing.IndexRoutingTests#testRoutingPathArraysInSource`.